art/defences/trainer/dp_instahide_trainer.py

```diff
@@ -0,0 +1,285 @@
+        self.noise = noise
+        self.loc = loc
+        self.scale = scale
+        self.clip_values = clip_values
+
+    def _generate_noise(self, x: np.ndarray) -> np.ndarray:
+        if self.noise == "gaussian":
+            noise = np.random.normal(loc=self.loc, scale=self.scale, size=x.shape)
+        elif self.noise == "laplacian":
+            noise = np.random.laplace(loc=self.loc, scale=self.scale, size=x.shape)
+        elif self.noise == "exponential":
+            noise = np.random.exponential(scale=self.scale, size=x.shape)
+        else:
+            raise ValueError("The provided noise type is not supported:", self.noise)
+
+        x_noise = x + noise
+        x_noise = np.clip(x_noise, self.clip_values[0], self.clip_values[1])
+
+        return x_noise
+
+    def fit(  # pylint: disable=W0221
+        self,
+        x: np.ndarray,
+        y: np.ndarray,
+        validation_data: Optional[Tuple[np.ndarray, np.ndarray]] = None,
+        batch_size: int = 128,
+        nb_epochs: int = 20,
+        **kwargs
+    ):
+        """
+        Train a model adversarially with the DP-InstaHide protocol.
+        See class documentation for more information on the exact procedure.
+
+        :param x: Training set.
+        :param y: Labels for the training set.
+        :param validation_data: Tuple consisting of validation data, (x_val, y_val)
+        :param batch_size: Size of batches.
+        :param nb_epochs: Number of epochs to use for trainings.
+        :param kwargs: Dictionary of framework-specific arguments. These will be passed as such to the `fit` function
+               of the target classifier.
+        """
+        logger.info("Performing adversarial training with DP-InstaHide protocol")
+
+        nb_batches = int(np.ceil(len(x) / batch_size))
+        ind = np.arange(len(x))
+
+        logger.info("Adversarial Training DP-InstaHide")
+
+        for i_epoch in trange(nb_epochs, desc="DP-InstaHide training epochs"):
+            # shuffle the examples
+            np.random.shuffle(ind)
+            start_time = time.time()
+            train_loss = 0.0
+            train_acc = 0.0
+            train_n = 0.0
+
+            for batch_id in range(nb_batches):
+                # create batch data
+                x_batch = x[ind[batch_id * batch_size : min((batch_id + 1) * batch_size, x.shape[0])]]
+                y_batch = y[ind[batch_id * batch_size : min((batch_id + 1) * batch_size, x.shape[0])]]
+
+                # augment the data
+                x_aug = x_batch.copy()
+                y_aug = y_batch.copy()
+                for augmentation in self.augmentations:
+                    x_aug, y_aug = augmentation(x_aug, y_aug)
+
+                # apply additive noise
+                x_aug = self._generate_noise(x_aug)
+
+                # extract label reduction and set to no reduction if needed
+                reduce_labels = self._classifier._reduce_labels  # type: ignore # pylint: disable=W0212
```

Comment:
Why is there a type ignore?

Reply:
Not all classifiers of type `CLASSIFIER_LOSS_GRADIENTS_TYPE` have the `_reduce_labels` property which causes a mypy error. This line is needed due to the `PyTorchClassifier` always reducing labels for `nn.CrossEntropyLoss`.